driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -387,21 +436,71 @@ private void maybeSpawnNewConnection() {
                 break;
         }
 
-        logger.debug("Creating new connection on busy pool to {}", host);
         manager.blockingExecutor().submit(newConnectionTask);
     }
 
     void replaceDefunctConnection(final PooledConnection connection) {
-        if (connection.markForTrash.compareAndSet(false, true))
+        if (connection.state.compareAndSet(OPEN, GONE))
             open.decrementAndGet();
-        connections.remove(connection);
+        if (connections.remove(connection))
+            manager.blockingExecutor().submit(new Runnable() {
+                @Override
+                public void run() {
+                    addConnectionIfUnderMaximum();
+                }
+            });
         connection.closeAsync();
-        manager.blockingExecutor().submit(new Runnable() {
-            @Override
-            public void run() {
-                addConnectionIfUnderMaximum();
+    }
+
+    void cleanupIdleConnections(long now) {
+        if (isClosed())
+            return;
+
+        shrinkIfBelowCapacity();
+        cleanupTrash(now);
+    }
+
+    /** If we have more active connections than needed, trash some of them */
+    private void shrinkIfBelowCapacity() {
+        int currentLoad = maxTotalInFlight.getAndSet(0);
+
+        int needed = currentLoad / StreamIdGenerator.MAX_STREAM_PER_CONNECTION + 1;
+        if (currentLoad % StreamIdGenerator.MAX_STREAM_PER_CONNECTION > options().getMaxSimultaneousRequestsPerConnectionThreshold(hostDistance))
+            needed += 1;
+        needed = Math.max(needed, options().getCoreConnectionsPerHost(hostDistance));
+        int actual = open.get();
+        int toTrash = Math.max(0, actual - needed);
+
+        logger.trace("Current inFlight = {}, {} connections needed, {} connections available, trashing {}",
+            currentLoad, needed, actual, toTrash);
+
+        if (toTrash <= 0)
+            return;
+
+        for (PooledConnection connection : connections)
+            if (trashConnection(connection)) {
```

Comment:
> Since connections is a list, it will always trash the oldest connections.

Yes. It's much easier to pick from the start of the list in a thread-safe way. But anyway which connections we trash doesn't matter, the incoming load will even out on the remaining ones.

> totalInFlight takes into account inFlight requests on trashed connections as well

Very good point. That's wrong and we should subtract a connection's inFlight from the total when it gets trashed.